File: sicp_1_8.js
#!/usr/bin/env node
// sicp exercise 1.8

function abs(x) {
  return x > 0 ? x : x < 0 ? -x : 0
}

function cube(x) { return x * x * x }

function goodEnough(guess,x) {
  return (abs((cube(guess)-x)) < 1e-100)
}

function improve(guess,x) {
  return (((x/(guess * guess)) + (2 * guess)) / 3.0)
}

function cubeRootIter(guess,x) {
  return goodEnough(guess,x) ? guess : cubeRootIter(improve(guess,x),x)
}

function cubeRoot(x) { return cubeRootIter(1.0,x) }

console.log(cubeRoot(27))
console.log(cubeRoot(64))

File: sicp_1_8.py
#!/usr/bin/env python

def abs_(x):
  if x < 0:
      return -x
  else:
      return x


def cube(x):
    return x * x * x


def iter_improve(update, test, guess=1):
    while not test(guess):
        guess = update(guess)
    return guess


def cube_root(a):
    return find_root(lambda x: cube(x) - a)


def logarithm(a, base=2):
    return find_root(lambda x: pow(base, x) - a)


def approx_derivative(f, x, delta=1e-5):
    df = f(x + delta) - f(x)
    return df/delta


def approx_eq(x, y, tolerance=1e-5):
    return abs_(x - y) < tolerance


def newton_update(f):
    def update(x):
      return x - f(x) / approx_derivative(f, x)
    return update


def find_root(f, initial_guess=10):
    def test(x):
        return approx_eq(f(x), 0)
    return iter_improve(newton_update(f), test, initial_guess)


print(cube_root(27))
print(cube_root(8))